R/gnps.R

```diff
@@ -0,0 +1,167 @@
+#' in the references.
+#'
+#' @param ppm for `join_gnps`: `numeric(1)` defining a relative, m/z-dependent,
+#'     maximal accepted difference between m/z values of peaks from the two
+#'     spectra to be matched/mapped.
+#'
+#' @param tolerance for `join_gnps`: `numeric(1)` defining a constant maximal
+#'     accepted difference between m/z values of peaks from the two spectra to
+#'     be matched/mapped.
+#'
+#' @param type for `join_gnps`: `character(1)` specifying the type of join that
+#'     should be performed. See [join()] for details and options. Defaults to
+#'     `type = "outer"`.
+#'
+#' @param x for `join_gnps`: `numeric` with m/z values from a spectrum. For
+#'     `gnps`: `matrix` with two columns `"mz"` and `"intensity"` containing
+#'     the peaks **aligned** with peaks in `y` (with `join_gnps`).
+#'
+#' @param xPrecursorMz for `join_gnps`: `numeric(1)` with the precursor m/z
+#'     of the spectrum `x`.
+#'
+#' @param y for `join_gnps`: `numeric` with m/z values from a spectrum. For
+#'     `gnps`: `matrix` with two columns `"mz"` and `"intensity"` containing
+#'     the peaks **aligned** with peaks in `x` (with `join_gnps`).
+#'
+#' @param yPrecursorMz for `join_gnps`: `numeric(1)` with the precursor m/z
+#'     of the spectrum `y`.
+#'
+#' @param ... for `join_gnps`: optional parameters passed to the [join()]
+#'     function. For `gnps`: ignored.
+#'
+#' @author Johannes Rainer, Michael Witting, based on the code from
+#'     Xing et al. (2020).
+#'
+#' @importFrom clue solve_LSAP
+#'
+#' @importFrom stats complete.cases
+#'
+#' @references
+#'
+#' Xing S, Hu Y, Yin Z, Liu M, Tang X, Fang M, Huan T. Retrieving and Utilizing
+#' Hypothetical Neutral Losses from Tandem Mass Spectra for Spectral Similarity
+#' Analysis and Unknown Metabolite Annotation. *Anal Chem.*
+#' 2020 Nov 3;92(21):14476-14483. \doi{10.1021/acs.analchem.0c02521}.
+#'
+#' @family grouping/matching functions
+#'
+#' @family distance/similarity functions
+#'
+#' @return
+#'
+#' See function definition in the description section.
+#'
+#' @rdname gnps
+#'
+#' @export
+#'
+#' @examples
+#'
+#' ## Define spectra
+#' x <- cbind(mz = c(10, 36, 63, 91, 93), intensity = c(14, 15, 999, 650, 1))
+#' y <- cbind(mz = c(10, 12, 50, 63, 105), intensity = c(35, 5, 16, 999, 450))
+#' ## The precursor m/z
+#' pmz_x <- 91
+#' pmz_y <- 105
+#'
+#' ## Plain join identifies only 2 matching peaks
+#' join(x[, 1], y[, 1])
+#'
+#' ## join_gnps finds 4 matches
+#' join_gnps(x[, 1], y[, 1], pmz_x, pmz_y)
+#'
+#' ## with one of the two precursor m/z being NA, the result are the same as
+#' ## with join.
+#' join_gnps(x[, 1], y[, 1], pmz_x, yPrecursorMz = NA)
+#'
+#' ## Calculate GNPS similarity score:
+#' map <- join_gnps(x[, 1], y[, 1], pmz_x, pmz_y)
+#' gnps(x[map[[1]], ], y[map[[2]], ])
+gnps <- function(x, y, ...) {
+    if (nrow(x) != nrow(y))
+        stop("'x' and 'y' are expected to be aligned peak matrices (i.e. ",
```

Comment:
The test here seems to test the equality of rows only, and I would thus expect the error message to complain about that. Would there be another test regarding the aligned nature of the matrices?

Reply:
OK, I will change the error message. I don't see a way how we could check that the peak matrices are indeed aligned.